File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.cpp
class Solution {
public:
set<string>ans;
void solve(int n,string & s)
{
    if(n==0 && !s.empty())
    {
        ans.insert(s);
        return ;
    }
    if(n==0)
    {
        return ;
    }
    if(s.empty())
    {  s.push_back('a');
        solve(n-1,s);
        s.pop_back();
         s.push_back('b');
        solve(n-1,s);
        s.pop_back();
         s.push_back('c');
        solve(n-1,s);
        s.pop_back();
    }
    else if(s.back()=='a')
    {   s.push_back('b');
        solve(n-1,s);
        s.pop_back();
        s.push_back('c');
        solve(n-1,s);
        s.pop_back();
    }
    else if(s.back()=='b')
    {   s.push_back('a');
        solve(n-1,s);
        s.pop_back();
        s.push_back('c');
        solve(n-1,s);
        s.pop_back();
    }
    else
    {  
        s.push_back('a');
        solve(n-1,s);
        s.pop_back();
        s.push_back('b');
        solve(n-1,s);
        s.pop_back();
    }
}
    string getHappyString(int n, int k) {
        string s;
      solve(n,s); 
      int i=1;
      for(string itr:ans)
      {
        if(i==k)
        {
            return itr;
        }
        i++;
      } 
      return "";
    }
};

File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.java
class Solution {
    int k;
   void solve(int n, StringBuilder st,ArrayList<String> ls)
   { 
    if(ls.size()==k)
   {
    return ;
   }
    if(n==0)
    {
        ls.add(st.toString());
        return;
    }
    if(st.length()==0)
    {
        st.append('a');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
         st.append('b');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
         st.append('c');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
    }
    else
    {
        char x=st.charAt(st.length()-1);
        if(x=='a')
        {
             st.append('b');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
         st.append('c');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
        }
          if(x=='b')
        {
             st.append('a');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
         st.append('c');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
        }
          if(x=='c')
        {
             st.append('a');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
         st.append('b');
        solve(n-1,st,ls);
        st.deleteCharAt(st.length()-1);
        }
    }
   }
    public String getHappyString(int n, int kk) {
        k=kk;
        ArrayList<String>ls=new ArrayList<>();
        StringBuilder st=new StringBuilder();
        solve(n,st,ls); 
        if(ls.size()<k)
        {
            return "";
        }
        String res=ls.get(k-1);
        return res;
    }
}